### ISSUE TITLE
Scanning is slower in 2.12.0

### ISSUE THREAD

**Issue by natros** [lukehutch/fast-classpath-scanner#174]
When scanning with 2.11.1 it takes 150ms and with 2.12.0 takes 450ms.

[2.11.1.txt](https://github.com/lukehutch/fast-classpath-scanner/files/1663789/2.11.1.txt)
[2.12.0.txt](https://github.com/lukehutch/fast-classpath-scanner/files/1663790/2.12.0.txt)

**Comment by lukehutch** [comment]
@natros Thanks for the report. I can't replicate the slowdown, on a couple of different large classpaths.

I noticed from your logfile that there was a bug in 2.11.1 that caused `java.class.path` to not be split properly on path separator characters `':'`, so relative to the 2.12.0 log, 2.11.1 missed a couple of jars that it was supposed to find (that weren't found through ClassLoaders), and so 2.12.0 did more scanning work than 2.11.1. Specifically, 2.12.0 scanned these additional jars:

```
/Users/fsousa/Library/Caches/IntelliJIdea2017.3/captureAgent/debugger-agent.jar
/Users/fsousa/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/173.4548.10/IntelliJ IDEA 2017.3 EAP.app/Contents/lib/asm-all.jar
```

Unless these jars were huge though, this wouldn't exchange the scanning time being tripled.

Is there a chance you could please attach your pom.xml file so I can replicate the problem with the same classpath?

**Comment by lukehutch** [comment]
PS I also profiled this extensively in JProfiler, and couldn't see a huge performance difference in any specific calls in version 2.12.0 compared to 2.11.1.

**Comment by natros** [comment]
Tested with JMH

**2.11.1**
```
Result "pt.ipb.creditacao.server.Scanner.scan":
  24.448 ±(99.9%) 0.111 ops/s [Average]
  (min, avg, max) = (21.388, 24.448, 25.450), stdev = 0.468
  CI (99.9%): [24.337, 24.558] (assumes normal distribution)


# Run complete. Total time: 00:07:01

Benchmark      Mode  Cnt   Score   Error  Units
Scanner.scan  thrpt  200  24.448 ± 0.111  ops/s
```

**2.12.0**
```
Result "pt.ipb.creditacao.server.Scanner.scan":
  11.467 ±(99.9%) 0.063 ops/s [Average]
  (min, avg, max) = (9.887, 11.467, 11.887), stdev = 0.265
  CI (99.9%): [11.405, 11.530] (assumes normal distribution)


# Run complete. Total time: 00:07:10

Benchmark      Mode  Cnt   Score   Error  Units
Scanner.scan  thrpt  200  11.467 ± 0.063  ops/s
```

**2.12.1-SNAPSHOT**
```
Result "pt.ipb.creditacao.server.Scanner.scan":
  10.249 ±(99.9%) 0.055 ops/s [Average]
  (min, avg, max) = (9.512, 10.249, 10.675), stdev = 0.234
  CI (99.9%): [10.193, 10.304] (assumes normal distribution)


# Run complete. Total time: 00:07:15

Benchmark      Mode  Cnt   Score   Error  Units
Scanner.scan  thrpt  200  10.249 ± 0.055  ops/s
```

**Comment by natros** [comment]
Test with a simple loop:

```java
public class Test {
  public static void main(String[] args) {
    System.out.println(FastClasspathScanner.getVersion());
    for (int i = 0; i < 10; i++) {
      Stopwatch timer = Stopwatch.createStarted();
      new FastClasspathScanner(Scanner.class.getPackage().getName()).scan();
      System.out.println("time=" + timer.stop().elapsed(TimeUnit.MILLISECONDS) + "ms");
    }
  }
}
```

**2.11.1**
```
time=223ms
time=90ms
time=72ms
time=89ms
time=77ms
time=86ms
time=75ms
time=73ms
time=61ms
time=87ms
```

**2.12.0**
```
time=459ms
time=599ms
time=192ms
time=181ms
time=151ms
time=142ms
time=149ms
time=151ms
time=117ms
time=127ms

```
**2.12.1-SNAPSHOT**
```
time=479ms
time=386ms
time=291ms
time=173ms
time=156ms
time=159ms
time=150ms
time=125ms
time=116ms
time=116ms
```

**Comment by natros** [comment]
This is only a stripped version
[pom.xml.gz](https://github.com/lukehutch/fast-classpath-scanner/files/1665969/pom.xml.gz)

**Comment by lukehutch** [comment]
@natros thanks so much for the detailed timing, and for the test project. I'm traveling for the next 3 weeks, but I will try to investigate soon. What OS and JVM / runtime environment are you running on?

**Comment by natros** [comment]
no rush, take your time.

java -version
java version "1.8.0_162"
Java(TM) SE Runtime Environment (build 1.8.0_162-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.162-b12, mixed mode)

macOS 10.13.3
Intellij IDEA 2017.3.4

**Comment by lukehutch** [comment]
@natros sorry for the delay on this. I think I may have found what was causing the slowdown. Can you please test 2.19.0? I will be pushing it out shortly. Thank you!

**Comment by lukehutch** [comment]
@natros I have gone in to profile this performance regression several times, with two different profiler tools, and couldn't figure out the problem until now. I looked at detailed profiling information and code diffs, couldn't find the source of the problem until now, since the profiling information wasn't telling me the right place to look (the timing values for the offending operation were being not properly captured, or something).

Good news though, I finally figured out what was causing the regression. I added some code that skipped ZipSFX headers if present, initially for #171 -- this required each zipfile to be opened twice (once as a byte stream, to check for the position of the "PK" header, and once az a `ZipFile`. My own testing and profiling indicated a negligible performance penalty for doing so, but it was affecting scan times quite a lot.

Bug #171 turned out to be solvable without skipping ZipSFX headers in the end, so I have disabled the ZipSFX skipping code by default.

In even better news, profiling the code helped me find another major performance problem -- it turns out `File#toURI()#toURL()` is a very expensive operation (taking ~10% of overall scanning time). I deferred this work until the point where the user requests URLs, so that most of the time this work is not done. So FCS is actually even faster now than it was in version 2.11.1.

The fixed code will be released in version 2.20.0. Sorry this took so long to resolve. Thank you for this bug report, and thank you for your patience!
